src/Features/Core/Portable/InheritanceMargin/Finders/BaseTypeSymbolsFinder.cs

```diff
@@ -0,0 +1,158 @@
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.InheritanceMargin.Finders
+{
+    internal class BaseTypeSymbolsFinder : InheritanceSymbolsFinder
+    {
+        public static readonly BaseTypeSymbolsFinder Instance = new();
+
+        /// <summary>
+        /// Return all the base interfaces and base types of <param name="symbol"/> in topological order.
+        /// e.g
+        /// 'class A : B { }'
+        /// 'class B : IB { }'
+        /// 'interface IB : IC { }'
+        /// If 'class A' is the input symbol, the result should be in the order like: 'B', 'IB', 'IC'.
+        /// </summary>
+        protected override Task<ImmutableArray<ISymbol>> GetAssociatedSymbolsAsync(
+            ISymbol symbol,
+            Solution solution,
+            CancellationToken cancellationToken)
+        {
+            var namedTypeSymbol = (INamedTypeSymbol)symbol;
+            if (namedTypeSymbol.BaseType == null)
+            {
+                // AllInterfaces are topologically sorted by default.
+                return Task.FromResult(namedTypeSymbol.AllInterfaces.CastArray<ISymbol>());
+            }
+
+            // If the symbol has base type, it means we need to topologically sort all the base interfaces and base types.
+            // Consider the all the base types and base interfaces as vertices, and each of them is pointed by its derived types/interfaces in the graph.
+            // We need an 'incomingSymbols' map, whose key is the vertice of the graph,
+            // the values is its derived types and interfaces to perform topologically sort.
+            // e.g.
+            // interface IA { }
+            // interface IB { }
+            // class A : IA { }
+            // class B : A, IB, IA { }
+            // 
+            // The map would be
+            // {
+            //     "IA": ["A", "B"],
+            //     "IB": ["B"],
+            //     "A": ["B"],
+            //     "B": []
+            // }
+            using var _ = GetPooledHashSetDictionary(out var incomingSymbolsMapBuilder);
+            var baseTypes = BaseTypeFinder.FindBaseTypes(namedTypeSymbol);
+
+            // Add the entry for all base types.
+            for (var i = 0; i < baseTypes.Length; i++)
+            {
+                // baseTypes are order like,
+                //                        [0]          [1]           [2]
+                // symbol_we_search -> baseClass1 -> baseClass2 -> baseClass3 ...
+                // There is no edge points to the 'baseClass1'.
+                // And since interface can't have base class, the item before a baseClass is just the previous item in the array.
+                // e.g. 'baseClass2' could only be pointed by 'baseClass1'.
+                var baseType = baseTypes[i];
+                var incomingSymbolsSetForBaseType = s_symbolHashSetPool.Allocate();
```

Comment:
All Hashset usage of `s_symbolHashSetPool.Allocate()` is disposed in the `using var _ = GetPooledHashSetDictionary(out var incomingSymbolsMapBuilder);` (See `SymbolSetDictionaryDisposer` below)
If this pattern is not easy to understand I am also fine to use the traddtional try-finally block to free the PooledDicionary and the PooledHashSet in it